### DISCUSSION TITLE
How to validate a signature is coming from the right address

### DISCUSSION THREAD

**Discussion by METRAVERSE** [aptos-labs/aptos-developer-discussions#312] [Questions]
### Describe your question in detail.

I'm verifying the signature on a backend and also sending the address from the front end, I need to validate the signature/public key is coming from the right address.

### Which operating system are you using?

Windows

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by 0x-j** [comment]
Hi I don't know if we have a function in sdk to do that, but can you refer to this code? 
https://github.com/aptos-labs/aptos-wallet-adapter/blob/bb1c3fd5974d9d4f5697003076c4be307702cd42/packages/wallet-adapter-core/src/LegacyWalletPlugins/WalletCoreV1.ts#L181

**Reply by METRAVERSE** [reply]
> Hi I don't know if we have a function in sdk to do that, but can you refer to this code? https://github.com/aptos-labs/aptos-wallet-adapter/blob/bb1c3fd5974d9d4f5697003076c4be307702cd42/packages/wallet-adapter-core/src/LegacyWalletPlugins/WalletCoreV1.ts#L181

I can't find anything in this function that double checks the wallet address matches the signature
I'm currently doing signature verification the same as it's done here
I just need to verify the wallet address being sent by the user matches the signature they provided, otherwise users could send a different address and a correct signature/message/public key and still sign in

**Reply by 0x-j** [reply]
i think the function i sent checks the signature matches the public key. can you make the frontend pass the public key to backend then let backend use the code i sent to check?

**Reply by 0x-j** [reply]
Maybe take a look at these 2 threads to see if there's anything useful. 
https://github.com/aptos-labs/aptos-developer-discussions/discussions/180
https://github.com/aptos-labs/aptos-developer-discussions/discussions/13

**Reply by METRAVERSE** [reply]
> i think the function i sent checks the signature matches the public key. can you make the frontend pass the public key to backend then let backend use the code i sent to check?

I'm checking the public key as is done there
but I don't know how to verify the address matches the public key
like validate that the public key is the public key for the address

**Reply by 0x-j** [reply]
on the frontend, if you use wallet adaptor you can get public key from the account

```ts
import { useWallet } from "@aptos-labs/wallet-adapter-react";
const { account } = useWallet();
const userPubKey = account.publicKey;
```

With this code you know that's the public key that paired with the address, cause address also come from same account.

**Reply by METRAVERSE** [reply]
> on the frontend, if you use wallet adaptor you can get public key from the account
> 
> ```ts
> import { useWallet } from "@aptos-labs/wallet-adapter-react";
> const { account } = useWallet();
> const userPubKey = account.publicKey;
> ```
> 
> With this code you know that's the public key that paired with the address, cause address also come from same account.

I'm doing this, I just mean I specifically need the address on the backend, and I need to verify the address is the same as the one used for the signature

**Reply by METRAVERSE** [reply]
@0xaptosj do you have any info about how the address is checked for matching the public key in the Blockchain? maybe I can copy code from there

I'm looking for something like this https://stackoverflow.com/questions/78024648/can-i-reverse-resolve-the-public-key-from-an-address-or-signature
resolving the public key from the address or something 
I know keys can be rotated so I'm not sure how it would work

**Reply by METRAVERSE** [reply]
@0xaptosj 
it seems very hard to validate this considering the implications of rotating the key 
https://github.com/aptos-labs/aptos-core/issues/566

can I use indexer to check the public key on the Blockchain?

**Reply by METRAVERSE** [reply]
I think I might have figured it out, I can get the authentication key of the address using indexer
and then I can get the authentication key from the public key from the front end and compare it

**Reply by 0x-j** [reply]
why can't you send public key from frontend to backend and just verify the tx is signed by the public key?  address is some hash of public key, so we just need to verify one of them i think. 

https://github.com/aptos-labs/aptos-developer-discussions/discussions/14#discussioncomment-8686874

**Reply by METRAVERSE** [reply]
> why can't you send public key from frontend to backend and just verify the tx is signed by the public key? address is some hash of public key, so we just need to verify one of them i think.
> 
> [#14 (comment)](https://github.com/aptos-labs/aptos-developer-discussions/discussions/14#discussioncomment-8686874)

I told you I'm already doing that, the problem is I need the address as well
I'm using the address as a user id, so I have to verify the address matches the wallet used for the signature

**Reply by 0x-j** [reply]
i don't think you need to verify address when you already verify public key
